File: WondayWall.iOS/WondayWall/Models/HistoryItem.swift
import Foundation

// 生成ステータス（定義順は処理の昇順）
enum GenerationStatus: String, Codable {
    case generating               = "generating"                    // 開始・プロンプト生成前（後方互換）
    case generatingPromptReady    = "generating_prompt_ready"       // プロンプト生成後・画像 API 前
    case generatingImageRequested = "generating_image_requested"    // 画像 API 呼び出し済み（再開不可）
    case success                  = "success"                       // 生成成功
    case skipped                  = "skipped"                       // スキップ（変化なし）
    case failure                  = "failure"                       // 生成失敗
}

// 生成履歴の1件
struct HistoryItem: Codable, Identifiable {
    var id: UUID
    let executedAt: Date
    let status: GenerationStatus
    // 使用したカレンダーイベント
    let usedCalendarEvents: [CalendarEventItem]?
    // 使用したニューストピック
    let usedNewsTopics: [NewsTopicItem]?
    // 使用したユーザープロンプト
    let usedPrompt: String?
    // エラー概要（失敗時のみ）
    let errorSummary: String?
    // 写真ライブラリに保存したアセット識別子（アルバム管理用）
    let photoAssetId: String?
    // テキストモデルが生成した画像プロンプト（generatingPromptReady 以降で設定、再開用）
    let generatedPrompt: String?

    var isSuccess: Bool { status == .success }
    var isSkipped: Bool { status == .skipped }
    var isGenerating: Bool {
        switch status {
        case .generating, .generatingPromptReady, .generatingImageRequested: return true
        default: return false
        }
    }

    init(
        id: UUID = UUID(),
        executedAt: Date = Date(),
        status: GenerationStatus,
        usedCalendarEvents: [CalendarEventItem]? = nil,
        usedNewsTopics: [NewsTopicItem]? = nil,
        usedPrompt: String? = nil,
        errorSummary: String? = nil,
        photoAssetId: String? = nil,
        generatedPrompt: String? = nil
    ) {
        self.id = id
        self.executedAt = executedAt
        self.status = status
        self.usedCalendarEvents = usedCalendarEvents
        self.usedNewsTopics = usedNewsTopics
        self.usedPrompt = usedPrompt
        self.errorSummary = errorSummary
        self.photoAssetId = photoAssetId
        self.generatedPrompt = generatedPrompt
    }
}

File: WondayWall.Android/app/src/main/kotlin/com/studiofreesia/wondaywall/models/HistoryItem.kt
package com.studiofreesia.wondaywall.models

import kotlinx.datetime.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

// 生成状態。生成中の中間状態も履歴に保存し、プロセス中断後の復旧判定に使う。
@Serializable
enum class GenerationStatus {
    @SerialName("generating")
    Generating,

    @SerialName("generating_prompt_ready")
    GeneratingPromptReady,

    @SerialName("generating_image_requested")
    GeneratingImageRequested,

    @SerialName("success")
    Success,

    @SerialName("skipped")
    Skipped,

    @SerialName("failure")
    Failure,
}

// 生成履歴の1件
@Serializable
data class HistoryItem(
    // 一意ID
    val id: String = UUID.randomUUID().toString(),
    // 実行日時
    val executedAt: Instant,
    // 生成状態
    val status: GenerationStatus,
    // エラー概要（失敗時のみ）
    val errorSummary: String? = null,
    // 写真領域に保存した画像 URI
    val appliedImageUri: String? = null,
    // 生成に使用したカレンダーイベント
    val usedCalendarEvents: List<CalendarEventItem>? = null,
    // 生成に使用したニューストピック
    val usedNewsTopics: List<NewsTopicItem>? = null,
    // 使用したサービスティア
    val serviceTier: GoogleAiServiceTier,
    // 生成に使用したユーザープロンプト（Android固有）
    val usedPrompt: String? = null,
    // テキストモデルが生成した画像プロンプト（再開用）
    val generatedPrompt: String? = null,
) {
    val isSuccess: Boolean
        get() = status == GenerationStatus.Success

    val isSkipped: Boolean
        get() = status == GenerationStatus.Skipped

    val isGenerating: Boolean
        get() = when (status) {
            GenerationStatus.Generating,
            GenerationStatus.GeneratingPromptReady,
            GenerationStatus.GeneratingImageRequested,
            -> true
            GenerationStatus.Success,
            GenerationStatus.Skipped,
            GenerationStatus.Failure,
            -> false
        }

}
